### PR TITLE
Validate package reference settings

### PR DESCRIPTION
**Pull request by tom-englert**
#851: Verify correct settings of PrivateAssets

### CODE REVIEW

File: Fody.sln.DotSettings [hunk 1]
```diff
@@ -12,6 +13,7 @@
 	<s:String x:Key="/Default/CodeInspection/Highlighting/InspectionSeverities/=MergeCastWithTypeCheck/@EntryIndexedValue">ERROR</s:String>
 	<s:String x:Key="/Default/CodeInspection/Highlighting/InspectionSeverities/=PartialMethodWithSinglePart/@EntryIndexedValue">ERROR</s:String>
 	<s:String x:Key="/Default/CodeInspection/Highlighting/InspectionSeverities/=PartialTypeWithSinglePart/@EntryIndexedValue">ERROR</s:String>
+	<s:String x:Key="/Default/CodeInspection/Highlighting/InspectionSeverities/=PossibleNullReferenceException/@EntryIndexedValue">DO_NOT_SHOW</s:String>
```
Comment (ltrzesniewski) [inline Fody.sln.DotSettings:16]: Are you sure we want to ignore these kinds of errors? Same with the one above.

Comment (tom-englert) [reply Fody.sln.DotSettings:16]: Yes, because they conflict with nullable ref types - I get contrary warnings from R#

Comment (ltrzesniewski) [reply Fody.sln.DotSettings:16]: Ok, I see - the R# support isn't foolproof. It should get better in 2020.2 hopefully.

**Comment by tom-englert** [comment]
It was intended to catch all package references in the first shot, even those that might be conditional and not active at the moment. (however conditional package references are not a common use case.)

Having package references in directory.build.props is not a common use case and not supported by Nuget installer, so if you do this manually, you should really know what you are doing.

Passing all package reference information to the inner weaver would have meant a lot of overhead of serializing and deserializing everything, so I went for the simpler solution (KISS/YAGNE).
This will cover everything that the Nuget Installer supports.

**Comment by ltrzesniewski** [comment]
I'd think that using the info provided by MSBuild would be a simpler solution. 😉
If you think serialization is a problem, why not pass the `ShouldCleanReference` value from the inner weaver back to Fody, and then do the check outside of the inner weaver?

**Comment by ltrzesniewski** [comment]
Would you like me to try this alternative implementation?

**Comment by tom-englert** [comment]
@ltrzesniewski if you got some time, give it a try.

**Comment by ltrzesniewski** [comment]
Ok, I'll do that. Actually I think I'll pass the value of the `PrivateAssets` metadata string of the weaver package to the inner weaver, it makes more sense than passing the value of `ShouldCleanReference` back to Fody.

**Comment by ltrzesniewski** [comment]
I added a commit that changes the implementation, but didn't fully test it yet, hence the draft status.
I think you can still review it though @tom-englert.

**Comment by tom-englert** [comment]
@ltrzesniewski the unit tests are not too meaningful with your implementation, since the code is no longer isolated. 
However I manually tested it with the Integration solution, and it behaved as expected, so from my side this is OK.

**Comment by ltrzesniewski** [comment]
Thanks! I haven't had the time to polish the unit tests properly, I merely tried to adapt the test cases you wrote. I'll try to improve on that, I just need to find some time to do that.

**Comment by tom-englert** [comment]
The tests are fine, I think that's all you can do without emulating the whole build process

**Comment by ltrzesniewski** [comment]
I think I should add some tests for the part in `WeavingTask`. That code relies on the fact that `ITaskItem.GetMetadata` returns an empty string (as opposed to null) for a metadata item that doesn't exist. I'd like to capture that in a test, along with some other stuff like making sure the code still works fine when the `PackageReference` doesn't use the same case in the package name as the weaver. I just haven't had the time to write those tests yet. 😕

**Comment by ltrzesniewski** [comment]
Ok, I added the test I wanted.
I rebased the branch because of a merge conflict.

**Comment by tom-englert** [comment]
Looks good, from my POV we can release this. 
@SimonCropp any objections?

**Pull request merged**